### DISCUSSION TITLE
Light Shader Projected UVs and Structs

### DISCUSSION THREAD

**Discussion by Kolupsy** [bnpr/Malt#341] [Q&A] [answered]
So I am basically just discovering light shaders and how they might be useful and I was wondering
1. How to get the UVs as projected from a sun light. I noticed that for point and spot light the `LightShaderInput.light_uv` is pretty much behaving as expected but for sun lights the light UV seems influenced by the view direction. Is that intentional and is there a way to get UVs from a sun light that is view-independant
2. How do I 'unpack' the `Light` struct in a light shader? There does not seem to be a `Light` struct available in the structs menu and its individual components are also not shown on the node. The only way I have found to access the light data is to use the Inline Code node
![image](https://user-images.githubusercontent.com/88381426/175661179-506f6903-96b2-4cea-be84-95730f01badb.png)

**Comment by pragma37** [comment, marked as answer]
Light Shaders need an improvement pass.

The light uv is the literal shadowmap uv, so yeah, in the case of sun lights those are view dependent. (Although there seems to be a bug there)

That said, it probably makes more sense to set the sun uvs to behave in a more "texture-mapping" way, using a "look at" matrix.

Light and LitSurface structs have been marked as internal, so they either should be removed from the IO nodes or made public again.

For LitSurface I'd say it doesn't make much sense to keep them anymore since all the info exposed is either not relevant in the Light Shader or is/can be exposed in a better way (e.g. set the global surface normal).

For Light I'm not that sure, but I think the light color and a well behaved uv should cover 99.99% of needs. (And in case more info is required it probably would make more sense to use GLSL anyway)

**Reply by pragma37** [reply]
I've published a Light Shader redesign: #342
I have not merged it into the Dev branch yet.

Any Feedback?
